src/develop/pixelpipe_hb.h

```diff
@@ -125,8 +125,11 @@ typedef struct dt_dev_pixelpipe_t
   int output_imgid;
   // working?
   int processing;
-  // shutting down?
-  int shutdown;
+  // shutting down?  Needs to be volatile so that compiler doesn't optimize away memory accesses
+  //  (a fully theoretically-correct implementation would use an atomic variable for the cross-thread
+  //  communication, but we're just using it to send a signal that may not be checked for tens of
+  //  milliseconds anyway.)
+  volatile int shutdown;
```

Comment:
It doesn't need to be sequentially consistent; a relaxed store would work just fine here, as the actual locking of the datta for which it would be bad if access isn't synchronized is still handled by a mutex.  The shutdown flag is merely a "please finish up early and release your lock ASAP" request.  Besides, holding the mutex for the entire duration of the pixelpipe run is in fact safer than the previous code's approach of repeatedly unlocking and relocking the mutex to give other threads a chance to even set the shutdown request.

But to keep you happy, I've added (simulated via mutex) atomics.  It seems that dt is *not* being compiled with C11 atomics support....

Reply:
I wonder why C11 atomics aren't available. It should depend on compiler version, so which ones are failing?

It's not immediately obvious if relaxed store/read won't have any bad effects right now - while it's true that reading FALSE from shutdown means only that shutdown just won't work (I'm not sure if we consider that a problem), but reading TRUE on the other hand will create a failure. You have to very carefully review how this variable is set and reset and if sequencing is guaranteed with relaxed memory order. Code is complex as it is, I don't think it's wise to create such dependencies, especially since any future modification will require the same review. Another aspect is atomicity itself - but int is most likely atomic everywhere anyway. To summarize - just write correct code and don't have any of these problems. If you want it more optimized - use non-default memory order for atomic operations.

Reply:
GCC 8.4, __STDC_NO_ATOMIC__ is **not** defined, but stdatomic.h itself threw errors that _Atomic is not defined.

As for the shutdown variable: reset in the initialization, set by exactly one function, and reset by another one guarded by the same mutex that the main pixelpipe runs inside of.  There are also two unguarded reads inside wait loops with timeouts in develop.c (waiting for a hash to become available) to abort the wait.

I've never heard of any architecture where load/store of an aligned machine word is not atomic, but you may need memory fences for anything other than relaxed order (depending on what ordering guarantees the architecture makes).   Unaligned, if supported by the architecture, is typically non-atomic due to needing two memory accesses.

Reply:
These two unguarded reads indeed caught my eyes, but I haven't tracked the call tree to understand if it's a problem. Which, as I said, already means there is a problem - it's not obvious if they can get spurious aborts from previous (aborted) runs of pipeline.

Reply:
I guess atomics are much better supported in C++ than in C. Which is sad, given that they were introduced at the same time.